In [2]:
import pandas as pd
import numpy as np

## Prepare and load training and testing datasets

In [3]:
pd.options.display.max_colwidth = 150

train_set = pd.read_csv("Train_Dataset.csv")


test_set = pd.read_csv("Test_Dataset.csv")
train_set.head(50)

headline  \
0                                                                 supreme court votes 7-2 to legalize all worldly vices   
1                                                    hungover man horrified to learn he made dozens of plans last night   
2                                                     emily's list founder: women are the 'problem solvers' in congress   
3                                                                         send your kids back to school with confidence   
4                                                                             watch: experts talk pesticides and health   
5                                           james corden and the red hot chili peppers strip down for 'carpool karaoke'   
6                                                                                 u.s. dignity reserves nearly depleted   
7                                                                how to re-ignite the spark in your body, mind and soul   
8                                           report: there still time to convert to christianity before christmas starts   
9                                                                                         education reform and evidence   
10                                                                   kid honors grandpa's memory with solemn cannonball   
11                                                         ragnar from 'vikings' is going where? creator talks season 3   
12                                                                 pope died as he lived: propped up for public viewing   
13                                documentary about plymouth rock throws in some world war ii to keep people interested   
14                                                                               jews, muslims, hindus agree on chicken   
15                                                                                           the new new net neutrality   
16                                                 white house struggled with asian leaders names and countries at g-20   
17                                                                    see chris hemsworth in his 'ghostbusters' uniform   
18                                                                                           should you marry that guy?   
19                                                parents drop fake treating-you-like-an-adult act half-hour into visit   
20                                                                                excercise ball all the way over there   
21                                                    20 people found refuge in a famous paris bookstore during attacks   
22                                                     actress leaves porn past behind with new cinemax erotic thriller   
23                                        kylie jenner and tyga got way handsy in the club like no one was snapchatting   
24                                transgender lawmaker danica roem: trump shows there's 'no barrier' to getting elected   
25                                                                  hoodie monks' use hip hop to impart buddhist wisdom   
26                                          concerned parents demand removal of arsenic from periodic table of elements   
27        confused zoo officials awkwardly celebrate after endangered panda gives birth to healthy northern white rhino   
28                                          lauren graham just dropped a clue about those final 4 'gilmore girls' words   
29                                                                                hero dog fills out hospital paperwork   
30                    logo of smiling cartoon tooth holding brush inspires nothing but confidence in local oral surgeon   
31                                                                        matt shepard is a friend of mine,' and my son   
32             doctors restore ken burns' full-color

In [4]:
#Checking for na values
train_set.isnull().sum()

headline        0
is_sarcastic    0
dtype: int64

In [5]:
train_set['is_sarcastic'].value_counts()
train_set.shape

(44262, 2)

In [6]:
test_set.shape

(11066, 1)

This means that about 54% of the data is non-sarcastic and 46% sarcastic.

## Split datasets

In [7]:
from sklearn.model_selection import train_test_split

X = train_set[['headline']]
y = train_set['is_sarcastic']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
X_train.shape

(29655, 1)

From the corpus, we will extract the following from each document:
* Word count - total number of words in the doc/headline
* Character count - total number of characters in the doc/headline
* Average word density - mean length of words in the doc

In [8]:
import string

X_train['char_count'] = X_train['headline'].apply(len)
X_train['word_count'] = X_train['headline'].apply(lambda x: len(x.split()))
X_train['word_density'] = X_train['char_count'] / (X_train['word_count']+1)

X_test['char_count'] = X_test['headline'].apply(len)
X_test['word_count'] = X_test['headline'].apply(lambda x: len(x.split()))
X_test['word_density'] = X_test['char_count'] / (X_test['word_count']+1)

# X_train['punctuation_count'] = X_train['headline'].apply(lambda x: len("".join(char for char in x if char in string.punctuation)))
# X_train['upper_case'] = X_train['headline'].apply(lambda x: len([word for word in x.split() if word.isupper()]))
# X_train['title_word_count'] = X_train['headline'].apply(lambda x: len([wrd for wrd in x.split() if wrd.istitle()]))

/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithC

In [9]:
X_train.head()

headline  \
22779                   google introduces a new way to screen telemarketers   
39644                              the true meaning of the ray rice scandal   
423                        a christmas message to vice president mike pence   
36581                               burger king unveils new low-fat cashier   
42039  argument about capital of australia occurs 10 feet from encyclopedia   

       char_count  word_count  word_density  
22779          51           8      5.666667  
39644          40           8      4.444444  
423            48           8      5.333333  
36581          39           6      5.571429  
42039          68          10      6.181818

In [10]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(random_state=42, solver='liblinear')

/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:35: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  eps=np.finfo(np.float).eps,
/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:597: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0

In [11]:
lr.fit(X_train.drop(['headline'], axis=1), y_train)
predictions = lr.predict(X_test.drop(['headline'], axis=1))

/Users/gbaldonado/opt/anaconda3/envs/dphi_bootcamp_nlp/lib/python3.7/site-packages/sklearn/linear_model/base.py:283: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  indices = (scores > 0).astype(np.int)


In [12]:
from sklearn import metrics

print(metrics.classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.58      0.75      0.65      7864
           1       0.55      0.36      0.44      6743

   micro avg       0.57      0.57      0.57     14607
   macro avg       0.57      0.56      0.55     14607
weighted avg       0.57      0.57      0.55     14607



In [13]:
print(metrics.confusion_matrix(y_test, predictions))

[[5878 1986]
 [4287 2456]]


In [14]:
print(metrics.accuracy_score(y_test,predictions))

0.5705483672211953


In [15]:
df = pd.DataFrame(metrics.confusion_matrix(y_test,predictions), index=['0','1'], columns=['0','1'])
df

0     1
0  5878  1986
1  4287  2456

<font color=green>These results are terrible. More non-sarcastic headlines are labeled as sarcastic (4287) than correctly identified as sarcastic (2456), although a relatively small number of sarcastic headlines (1986) were confused as non-sarcastic. Let's try another classifier called Naive Bayes.</font>

In [16]:
from sklearn.naive_bayes import MultinomialNB

nb = MultinomialNB()

nb.fit(X_train.drop(['headline'], axis=1), y_train)
predictions = nb.predict(X_test.drop(['headline'], axis=1))

In [17]:
print(metrics.confusion_matrix(y_test, predictions))

[[7862    2]
 [6692   51]]


In [18]:
print(metrics.classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.54      1.00      0.70      7864
           1       0.96      0.01      0.02      6743

   micro avg       0.54      0.54      0.54     14607
   macro avg       0.75      0.50      0.36     14607
weighted avg       0.74      0.54      0.38     14607



In [19]:
print(metrics.accuracy_score(y_test,predictions))

0.541726569453002


<font color=green>Clearly, the predictions using the naive bayes classifier is worst than the predictions of the logistic regression model. Although more non-sarcastic headlines are correctly predicted (7862), almost none (51) are correctly recognized as sarcastic. The false non-sarcastic headlines shot up to 6692 counts from 4287 from the logistic regression model. Let's try another model from the support vector machine (SVM) classifier.</font>

In [ ]:
from sklearn.svm import SVC
svc_model = SVC(gamma="auto")
svc_model.fit(X_train.drop(['headline'], axis=1), y_train)

In [ ]:
predictions = svc_model.predict(X_test.drop(['headline'], axis=1))

In [ ]:
print(metrics.confusion_matrix(y_test, predictions))

In [ ]:
print(metrics.classification_report(y_test, predictions))

In [ ]:
print(metrics.accuracy_score(y_test,predictions))

<font color=green>The SVC model performs the best compared to the logistic regression and naive bayes models. We will aggregate the results from this model to the results of sentiment analysis to get a better predictions</font>

## Text Pre-Processing and Wrangling

In [ ]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')

In [ ]:
import contractions
import re

stop_words = nltk.corpus.stopwords.words('english')

def normalize_doc(doc):
    # lower case and remove special characters\whitespaces
    doc = re.sub(r'[^a-zA-Z\s]', '', doc, re.I|re.A)
    doc = doc.lower()
    doc = doc.strip()
    # tokenize document
    tokens = nltk.word_tokenize(doc)
    # filter stopwords out of document
    filtered_tokens = [token for token in tokens if token not in stop_words]
    # re-create document from filtered tokens
    doc = ' '.join(filtered_tokens)
    return doc

normalize_corpus = np.vectorize(normalize_doc)

In [ ]:
X_train['clean headline'] = normalize_corpus(X_train['headline'].values)
X_test['clean headline'] = normalize_corpus(X_test['headline'].values)

In [ ]:
X_train.reset_index(drop=True, inplace=True)
X_train

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC

text_clf = Pipeline([('tfidf', TfidfVectorizer()),
                     ('clf', LinearSVC()),
])

# Feed the training data through the pipeline
text_clf.fit(X_train['headline'], y_train)  

In [ ]:
from sklearn import metrics

In [ ]:
predictions = text_clf.predict(X_test['headline'])

In [ ]:
print(metrics.classification_report(y_test, predictions))

In [ ]:
print(metrics.confusion_matrix(y_test, predictions))

In [ ]:
print(metrics.classification_report(y_test,predictions))

In [ ]:
results = text_clf.predict(test_set['headline'])

In [ ]:
test_set['predicted sarcasm'] = results

In [ ]:
test_set